composable-tasks/src/compose_tasks.rs

```diff
@@ -0,0 +1,106 @@
+use core::convert::TryFrom;
+
+use crate::external_sc_interactions;
+
+multiversx_sc::imports!();
+multiversx_sc::derive_imports!();
+
+#[derive(TypeAbi, TopEncode, TopDecode, PartialEq, ManagedVecItem)]
+pub enum TaskType {
+    WrapEGLD,
+    UnwrapEGLD,
+    Swap,
+    RouterSwap,
+    SendEgldOrEsdt,
+}
+
+#[multiversx_sc::module]
+pub trait TaskCall:
+    external_sc_interactions::pair_actions::PairActionsModule
+    + external_sc_interactions::router_actions::RouterActionsModule
+    + external_sc_interactions::wegld_swap::WegldWrapModule
+{
+    #[payable("*")]
+    #[endpoint(composeTasks)]
+    fn compose_tasks(
+        &self,
+        expected_token_out: EgldOrEsdtTokenPayment,
+        tasks: MultiValueEncoded<MultiValue2<TaskType, ManagedVec<ManagedBuffer>>>,
+    ) {
+        let payment = self.call_value().egld_or_single_esdt();
+        let mut payment_for_next_task = payment;
+
+        let caller = self.blockchain().get_caller();
+
+        for task in tasks.into_iter() {
+            let (task_type, args) = task.into_tuple();
+
+            let payment_for_current_task = payment_for_next_task.clone();
+
+            payment_for_next_task = match task_type {
+                TaskType::WrapEGLD => self.wrap_egld(payment_for_current_task),
+                TaskType::UnwrapEGLD => self.unwrap_egld(payment_for_current_task),
+                TaskType::Swap => {
+                    require!(
+                        !payment_for_current_task.token_identifier.is_egld(),
+                        "EGLD can't be swapped!"
```

Comment:
Not really sure we need this check. In the next line, when you do `unwrap_esdt` you would still get an error that you expect an ESDT.

Reply:
The check for EGLD is necessary because of the next instruction: unwrap_esdt. If someone provided EGLD by mistake we want to nicely output the error.